File: tensorrt/onnx_to_tensorrt/onnx2trt.py
import tensorrt as trt
import numpy as np
import common
import time
import argparse


def main():
    parser = argparse.ArgumentParser(description = 'Parameters for keras model convert to onnx')
    parser.add_argument('--onnx_path', '-onnx', required = True, type = str, 
    help = 'onnx model file path')
    parser.add_argument('--trt_path', '-trt', required = True, type = str, 
    help = 'save path of tensorrt model')
    args = parser.parse_args()

    model_path = args.onnx_path

    TRT_LOGGER = trt.Logger(trt.Logger.VERBOSE)
    EXPLICIT_BATCH = 1 << (int)(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
    builder = trt.Builder(TRT_LOGGER)
    network = builder.create_network(EXPLICIT_BATCH)
    parser = trt.OnnxParser(network, TRT_LOGGER)
    model = open(model_path, 'rb')

    if not parser.parse(model.read()):
        for error in range(parser.num_errors):
            print(parser.get_error(error))

    engine = builder.build_engine(network, builder.create_builder_config())
    
    plan=engine.serialize()
    with open(args.trt_path, "wb") as f:
        f.write(plan)
    f.close()

if __name__ == "__main__":
    main()

File: tensorrt/onnx_to_tensorrt/onnx2trt.cpp
#include <NvOnnxParser.h>
#include <NvInfer.h>
#include <NvInferRuntimeCommon.h>

#include <fstream>
#include <iostream>
#include <string>

using namespace std;

class Logger : public ILogger           
{
    void log(Severity severity, const char* msg) override
    {
        // suppress info-level messages
        if (severity != Severity::kINFO)
            std::cout << msg << std::endl;
    }
};

// refer to https://github.com/onnx/onnx-tensorrt/blob/84b5be1d6fc03564f2c0dba85a2ee75bad242c2e/main.cpp
int main()
{
    Logger logger;
    const auto explicitBatch = 1U << static_cast<uint32_t>(nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH);
    auto builder = nvinfer1::createInferBuilder(logger);  
    auto network = builder->createNetworkV2(explicitBatch);
    auto onnxparser  = nvonnxparser::createParser(*network, logger);

    string onnxfilename = "model.onnx";
    std::ifstream onnxfile(onnx_filename.c_str(),
                            std::ios::binary | std::ios::ate);
    std::streamsize filesize = onnxfile.tellg();
    onnxfile.seekg(0, std::ios::beg);
    std::vector<char> onnxbuf(file_size);
    if( !onnxfile.read(onnxbuf.data(), onnxbuf.size()) ) {
        cerr << "ERROR: Failed to read from file " << onnxfilename << endl;
        return -4;
    }

    auto success = onnxparser->parse(onnxbuf.data(), onnxbuf.size());
    if (!success) return -1;

    size_t max_batch_size = 1;
    size_t max_workspace_size = 1<<30;
    builder->setMaxBatchSize(max_batch_size);
    builder->setMaxWorkspaceSize(max_workspace_size);

    auto engine = builder->buildCudaEngine(*network.get());
    auto plan = engine->serialize();
    engine_filename = "model.trt"
    ofstream engine_file(engine_filename.c_str());
    if (!engine_file) {
      cerr << "Failed to open output file for writing: "
           << engine_filename << endl;
      return -1;
    }

    engine_file.write((char*)plan->data(), plan->size());
    engine_file.close();

    return 0;
}